apiserver/facades/agent/uniter/uniter.go

```diff
@@ -2447,16 +2447,36 @@ func (u *UniterAPI) SetPodSpec(args params.SetPodSpecParams) (params.ErrorResult
 	return results, nil
 }
 
+// GetCloudSpec returns the cloud spec used by the model in which the
+// authenticated unit or application resides.
+// A check is made beforehand to ensure that the request is made by an entity
+// that has been granted the appropriate trust.
+func (u *UniterAPI) CloudSpec() (params.CloudSpecResult, error) {
+	canAccess, err := u.accessCloudSpec()
+	if err != nil {
+		return params.CloudSpecResult{}, err
+	}
+	if !canAccess() {
+		return params.CloudSpecResult{Error: common.ServerError(common.ErrPerm)}, nil
+	}
+
+	return u.cloudSpec.GetCloudSpec(u.m.Tag().(names.ModelTag)), nil
+}
+
 // GoalStates returns information of charm units and relations.
-func (u *UniterAPI) GoalStates(args params.Entities) (string, error) {
-	result := params.StringResults{
-		Results: make([]params.StringResult, len(args.Entities)),
+func (u *UniterAPI) GoalStates(args params.Entities) (params.GoalStateResults, error) {
+	result := params.GoalStateResults{
+		Results: make([]params.GoalStateResult, len(args.Entities)),
 	}
+
 	canAccess, err := u.accessUnit()
 	if err != nil {
-		return "", err
+		logger.Debugf("can't access uniter API")
+		return params.GoalStateResults{}, err
 	}
+
 	for i, entity := range args.Entities {
+		logger.Debugf("application %v, entity %v ", entity.Tag, len(args.Entities))
```

Comment:
Again, Trace.

Reply:
I suspect this was added to aid development - I do think it can /should be deleted now